### PR TITLE
Loader and Trigger for EXTRACT

### PR DESCRIPTION
**Pull request by tdincer**
Key Functions:
- [x] compose a Matlab script to define an EXTRACT analysis
- [x] trigger EXTRACT analysis in Matlab from python.
- [x] load the EXTRACT analysis results.

### CODE REVIEW

File: element_interface/extract_loader.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        from scipy.io import loadmat
+
+        output_file = list(Path(extract_dir).glob("*_extract_output.mat"))
+
+        if not output_file:
+            raise FileNotFoundError(
+                "EXTRACT output result files not found at {}".format(extract_dir)
+            )
+
+        results = loadmat(output_file[0])
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        try:
            extract_file = next(Path(extract_dir).glob("*_extract_output.mat"))
        except StopInteration:
            raise FileNotFoundError(
                f"EXTRACT output result file is not found at {extract_dir}."
            )

        results = loadmat(extract_file)
```

File: element_interface/extract_loader.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from pathlib import Path
+from datetime import datetime
+
+
+class EXTRACT_loader:
+    def __init__(self, extract_dir: str):
+        """Initialize EXTRACT loader class
+
+        Args:
+            extract_dir (str): string, absolute file path to Extract directory
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
            extract_dir (str): string, absolute file path to EXTRACT directory
```

File: element_interface/extract_loader.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        output_file = list(Path(extract_dir).glob("*_extract_output.mat"))
+
+        if not output_file:
+            raise FileNotFoundError(
+                "EXTRACT output result files not found at {}".format(extract_dir)
+            )
+
+        results = loadmat(output_file[0])
+
+        self.creation_time = datetime.utcnow()
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        self.creation_time = datetime.fromtimestamp(os.stat(extract_file).st_ctime)
```

File: element_interface/extract_loader.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
```
Comment (kabilar) [inline element_interface/extract_loader.py:2]:
```suggestion
import os
import numpy as np
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,7 @@ Observes [Semantic Versioning](https://semver.org/spec/v2.0.0.html) standard and
 
 + Add - mkdocs documentation
 + Add - improved docstrings for mkdocs
++ Add - EXTRACT trigger and loader tools
```
Comment (kabilar) [inline CHANGELOG.md:9]: Please update the version here to `0.4.0`, update the `version.py` file, and add the following the end of this file:
```
[0.4.0]: https://github.com/datajoint/element-interface/releases/tag/0.4.0
```

File: element_interface/extract_loader.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        if not output_file:
+            raise FileNotFoundError(
+                "EXTRACT output result files not found at {}".format(extract_dir)
+            )
+
+        results = loadmat(output_file[0])
+
+        self.creation_time = datetime.utcnow()
+        self.S = results["output"][0]["spatial_weights"][0]  # (Height, Width, Time)
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        self.S = results["output"][0]["spatial_weights"][0]  # (Height, Width, MaskId)
```

File: element_interface/extract_loader.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        self.T = results["output"][0]["temporal_weights"][0]  # (Time, MaskId)
+
+    def load_results(self):
+        """Load the EXTRACT results"""
+        from scipy.sparse import find
+
+        S_tr = self.S.transpose([2, 0, 1])  # Mask_no, Height, Width
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        S_transposed = self.S.transpose([2, 0, 1])  # MaskId, Height, Width
```

Comment (kabilar) [reply element_interface/extract_loader.py]: I changed the variable name to improve readability according to PEP8.

File: element_interface/extract_loader.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            )
+
+        results = loadmat(output_file[0])
+
+        self.creation_time = datetime.utcnow()
+        self.S = results["output"][0]["spatial_weights"][0]  # (Height, Width, Time)
+        self.T = results["output"][0]["temporal_weights"][0]  # (Time, MaskId)
+
+    def load_results(self):
+        """Load the EXTRACT results"""
```
Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        """Load the EXTRACT results
        
        Returns:
            masks (dict): Details of the masks identified with the EXTRACT segmentation package.
        """
```

File: element_interface/extract_loader.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        from scipy.sparse import find
+
+        S_tr = self.S.transpose([2, 0, 1])  # Mask_no, Height, Width
+
+        masks = []
+
+        for mask_id, s in enumerate(S_tr):
```
Comment (kabilar) [inline element_interface/extract_loader.py:39]:
```suggestion
        for mask_id, s in enumerate(S_transposed):
```

Comment (kabilar) [inline element_interface/extract_loader.py]:
```suggestion
        for mask_id, s in enumerate(S_transposed):
```

File: element_interface/extract_loader.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                dict(
+                    mask_id=mask_id,
+                    mask_npix=len(weights),
+                    mask_weights=weights,
+                    mask_center_x=int(np.average(xpixels, weights=weights) + 0.5),
+                    mask_center_y=int(np.average(ypixels, weights=weights) + 0.5),
+                    mask_center_z=None,
+                    mask_xpix=xpixels,
+                    mask_ypix=ypixels,
+                    mask_zpix=None,
```
Comment (kabilar) [inline element_interface/extract_loader.py:56]: Should we update the attributes `mask_center_z` and `mask_zpix` to be nullable in the `imaging.Segmentation.Mask` table?

File: element_interface/extract_trigger.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        """A helper class to trigger EXTRACT analysis in element-calcium-imaging.
+
+        Args:
+            scanfile_fullpath (Union[str, Path]): Full path of the scan
+            parameters (dict): EXTRACT input paramaters.
+            output_dir (Union[str, Path]): Directory to store the outputs of EXTRACT analysis.
+        """
+        assert isinstance(parameters, dict)
+
+        self.scanfile = Path(scanfile_fullpath)
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
        self.scanfile = Path(scanfile)
```

Comment (kabilar) [reply element_interface/extract_trigger.py]: I would prefer not renaming in this case for improved readability.

File: element_interface/extract_trigger.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        
+        % Perform Extraction
+        output = extractor(M, config);
+        save('{output_fullpath}', 'output');
+        """
+    )
+
+    def __init__(
+        self,
+        scanfile_fullpath: Union[str, Path],
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
        scanfile: Union[str, Path],
```

File: element_interface/extract_trigger.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        parameters: dict,
+        output_dir: Union[str, Path],
+    ) -> None:
+        """A helper class to trigger EXTRACT analysis in element-calcium-imaging.
+
+        Args:
+            scanfile_fullpath (Union[str, Path]): Full path of the scan
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
            scanfile (Union[str, Path]): Full path of the scan
```

File: element_interface/extract_trigger.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            eng = matlab.engine.start_matlab()
+            eng.run_extract()
+        except Exception as e:
+            raise e
+        finally:
+            os.chdir(current_dir)
+
+        run_status["execution_duration"] = (
+            datetime.utcnow() - run_status["processing_time"]
+        ).total_seconds()
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
```

Comment (kabilar) [reply element_interface/extract_trigger.py]: Was this only used for debugging?

File: element_interface/extract_trigger.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+    def run(self):
+        """Run the matlab run_extract.m script."""
+
+        self.write_matlab_run_script()
+
+        current_dir = Path.cwd()
+        os.chdir(self.output_dir)
+
+        run_status = {"processing_time": datetime.utcnow()}
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
```

**Comment by kabilar** [comment]
Do we need to add the matlab engine to `requirements.txt` or just update the [Element Calcium Imaging tutorial docs](https://datajoint.com/docs/elements/element-calcium-imaging/0.4/tutorials/)?

File: element_interface/extract_loader.py (hunk 3 above)
Comment (tdincer) [reply element_interface/extract_loader.py]: What's the benefit of this?

File: element_interface/extract_loader.py (hunk 10 above)
Comment (tdincer) [reply element_interface/extract_loader.py:56]: I did.

File: CHANGELOG.md (hunk 5 above)
Comment (tdincer) [reply CHANGELOG.md:9]: I'll keep it at 0.3.1 To be consistent with the previous unreleased changes at 0.3.1

Comment (kabilar) [reply CHANGELOG.md:9]: We should make a release with these updates.  And these changes warrant a bump to `0.4.0`.

File: element_interface/extract_trigger.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+                    [
+                        f"config.{k} = '{v}';"
+                        if isinstance(v, str)
+                        else f"config.{k} = {str(v).lower()};"
+                        if isinstance(v, bool)
+                        else f"config.{k} = {v};"
+                        for k, v in self.parameters.items()
+                    ]
+                ),
+                scanfile=self.scanfile,
```
Comment (kabilar) [inline element_interface/extract_trigger.py]: Add `.as_posix()`?

File: element_interface/extract_trigger.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+        """
+        % Load Data
+        data = load('{scanfile}');
+        M = data.M;
+        
+        % Input Paramaters
+        config = struct();
+        {parameters_list_string}
+        
+        % Perform Extraction
```
Comment (kabilar) [inline element_interface/extract_trigger.py]:
```suggestion
        % Run EXTRACT
```

**Review by kabilar — CHANGES_REQUESTED** [review]
Thanks @tdincer!  Great work.  A few suggestions above.

**Review by kabilar — APPROVED** [review]
Thanks @tdincer!

**Pull request merged**